app/service/learning_svc.py

```diff
@@ -35,43 +36,60 @@ async def build_model(self):
 
-    async def learn(self, facts, link, blob):
+    async def learn(self, facts, link, blob, operation=None):
         decoded_blob = b64decode(blob).decode('utf-8')
 
         found_facts = []
         for parser in self.parsers:
             try:
                 for fact in parser.parse(decoded_blob):
-                    await self._save_fact(link, facts, fact)
                     found_facts.append(fact)
             except Exception as e:
                 self.log.error(e)
         await self._update_scores(link, facts, increment=len(found_facts))
-        await self._build_relationships(link, found_facts)
+        await self._build_relationships(link, found_facts, operation)
 
     """ PRIVATE """
 
     @staticmethod
     async def _update_scores(link, facts, increment):
+        knowledge_svc_handle = BaseService.get_service('knowledge_svc')
         for uf in link.facts:
             for found_fact in facts:
                 if found_fact.unique == uf.unique:
                     found_fact.score += increment
+                    knowledge_svc_handle.update_fact(dict(trait=found_fact.trait, value=found_fact.value,
+                                                          source=found_fact.source), dict(score=found_fact.score))
                     break
 
     @staticmethod
-    async def _save_fact(link, facts, fact):
-        if all(fact.trait) and not any(f.trait == fact.trait and f.value == fact.value for f in facts):
+    async def _save_fact(link, facts, fact, operation=None):
+        if all(fact.trait) and not any(f.trait == fact.trait and f.value == fact.value and f.source == link.id
+                                       for f in facts):
             fact.collected_by = link.paw
             fact.technique_id = link.ability.technique_id
+            fact.links = [link]
+            fact.relationships = []
+            fact.source_type = OriginType.LEARNED.name
+            fact.source = operation.id if operation else link.id
+            knowledge_svc_handle = BaseService.get_service('knowledge_svc')
+            await knowledge_svc_handle.add_fact(fact)
             link.facts.append(fact)
 
-    async def _build_relationships(self, link, facts):
+    async def _build_relationships(self, link, facts, operation=None):
+        facts_covered = []
         for relationship in self.model:
             matches = []
             for fact in facts:
                 if fact.trait in relationship:
                     matches.append(fact)
+                    facts_covered.append(fact)
+                else:
+                    await self._save_fact(link, facts, fact)
             for pair in itertools.combinations(matches, r=2):
                 if pair[0].trait != pair[1].trait:
-                    link.relationships.append(Relationship(source=pair[0], edge='has', target=pair[1]))
+                    await link._create_relationships([Relationship(source=pair[0], edge='has', target=pair[1])],
```

Comment:
So a new relationship between facts would only be created if existing relationships exist in the model? i.e. if no relationships, just skip to last block below and save the facts but build no relationships between them?

Reply:
Yep, that's pretty much it. Unfortunately, this isn't a super clean method, but this approach seemed to be the best option to make sure that relationships could still be formed based on the model, that facts generated via that relationship can reference said relationship (for tracking/analysis purposes), and that we wouldn't miss any facts if they didn't fit in a relationship (which actually happens pretty frequently if you don't configure a model, like some tests do).